assets/js/api.js

```diff
@@ -1,644 +1,368 @@
+          map.removeControl(timelineLayer.timeSliderControl);
+          timelineLayer.timeSliderControl = L.Timeline.timeSliderControl(timelineLayer);
+          timelineLayer.timeSliderControl.addTo(map);
+        }
+        timelineLayer.addTo(map);
+        $loading.hide();
       });
-
-      $("#sidebar input:checked").each(function() {
-        formVars.push($(this).val());
+    } else if (needPermits) {
+      $.when(
+        getPermits(yearStart + '-01-01', yearEnd + '-12-31', nbhoodVal, "residential", "permits")
+      ).then(function() {
+        if (typeof timelineLayer.timeSliderControl != "undefined") {
+          map.removeControl(timelineLayer.timeSliderControl);
+          timelineLayer.timeSliderControl = L.Timeline.timeSliderControl(timelineLayer);
+          timelineLayer.timeSliderControl.addTo(map);
+        }
+        timelineLayer.addTo(map);
+        $loading.hide();
       });
-      for (var i = 0; i < formVars.length; i++) {
-        switch (formVars[i]) {
-          case "permits":
-            // Restricting date selection to full year
-            var needPermits = true;
-            break;
-          case "crimes":
-            if (parseInt(yearStart) < 2004) {
-              yearStart = "2004";
-            }
-            var yearRange = yearsInRange(yearStart, yearEnd);
-            if (yearRange.length == 0) {
-              console.log('Year range is zero!');
-            }
-            $('#crimetotal').html('');
-            getCrimesYear(nbhoodVal, yearRange);
-            break;
-          case "demolitions":
-            var needDemolitions = true;
-            break;
-
+    } else if (needDemolitions) {
+      $.when(
+        getPermits(yearStart + '-01-01', yearEnd + '-12-31', nbhoodVal, "residential", "demolitions")
+      ).then(function() {
+        if (typeof timelineLayer.timeSliderControl != "undefined") {
+          map.removeControl(timelineLayer.timeSliderControl);
+          timelineLayer.timeSliderControl = L.Timeline.timeSliderControl(timelineLayer);
+          timelineLayer.timeSliderControl.addTo(map);
         }
-      }
-
-      // Should refactor this! Maybe attach a data attribute to each checkbox,
-      // then iterate through those needing to be added to the animation and
-      // call get permits for each of them, then apply 'then' callback
-      if (needPermits && needDemolitions) {
-        $.when(
-          // if we need to make sure permits finishes before demolitions b/c of the date range,
-          // can chain another "then" with the demolitions call; but we should instead constrain
-          // the date selection to the data availability
-          getPermits(yearStart + '-01-01', yearEnd + '-12-31', nbhoodVal, "residential", "permits"),
-          getPermits(yearStart + '-01-01', yearEnd + '-12-31', nbhoodVal, "residential", "demolitions")
-        ).then(function() {
-          if (typeof timelineLayer.timeSliderControl != "undefined") {
```

Comment:
These 9 lines of code defining the 'then' callback function, look the same for each of the 3 queries related to permits and demolitions, so that function can be defined once and then referenced in the 3 then clauses.